chemprop/utils.py

```diff
@@ -420,58 +420,89 @@ def wrap(*args, **kwargs) -> Any:
     """
     makedirs(save_dir)
 
     with open(data_path) as f:
         reader = csv.reader(f)
         header = next(reader)
 
-        if smiles_column is None:
-            smiles_column_index = 0
+        smiles_columns = preprocess_smiles_columns(smiles_columns)
+        if None in smiles_columns:
+            smiles_columns = header[:len(smiles_columns)]
+            smiles_columns_index = list(range(len(smiles_columns)))
         else:
-            smiles_column_index = header.index(smiles_column)
+            smiles_columns_index = [header.index(i) for i in smiles_columns]
 
-        lines_by_smiles = {}
         indices_by_smiles = {}
         for i, line in enumerate(reader):
-            smiles = line[smiles_column_index]
-            lines_by_smiles[smiles] = line
+            smiles = tuple(line[j] for j in smiles_columns_index)
             indices_by_smiles[smiles] = i
 
+    if task_names is None:
+        task_names = [i for i in header if i not in smiles_columns]
+
+    features_header = []
+    if features_path is not None:
+        for feat_path in features_path:
+            with open(feat_path, 'r') as f:
+                reader = csv.reader(f)
+                feat_header=next(reader)
+                features_header.extend(feat_header)
+
     all_split_indices = []
     for dataset, name in [(train_data, 'train'), (val_data, 'val'), (test_data, 'test')]:
         if dataset is None:
             continue
 
         with open(os.path.join(save_dir, f'{name}_smiles.csv'), 'w') as f:
             writer = csv.writer(f)
-            writer.writerow(['smiles'])
+            if smiles_columns[0]=='':
+                writer.writerow(['smiles'])
+            else:
+                writer.writerow(smiles_columns)
             for smiles in dataset.smiles():
-                writer.writerow([smiles])
+                writer.writerow(smiles)
 
         with open(os.path.join(save_dir, f'{name}_full.csv'), 'w') as f:
             writer = csv.writer(f)
-            writer.writerow(header)
-            for smiles in dataset.smiles():
-                writer.writerow(lines_by_smiles[smiles])
+            writer.writerow(smiles_columns+task_names)
+            dataset_targets=dataset.targets()
+            for i,smiles in enumerate(dataset.smiles()):
+                writer.writerow(smiles+dataset_targets[i])
+
+        dataset_features=dataset.features()
+        if features_path is not None:
+            with open(os.path.join(save_dir, f'{name}_features.csv'), 'w') as f:
+                writer = csv.writer(f)
+                writer.writerow(features_header)
+                writer.writerows(dataset_features)
 
         split_indices = []
         for smiles in dataset.smiles():
-            split_indices.append(indices_by_smiles[smiles])
+            try:
+                split_indices.append(indices_by_smiles[tuple(smiles)])
+            except KeyError:
+                split_indices.append(None)
```

Comment:
You could use the `get()` method here instead, which returns `None` by default if the key is not in the dictionary.
```suggestion
            split_indices.append(indices_by_smiles.get(tuple(smiles)))
```